### ISSUE TITLE
Vampires often shift back into their own form at low HP

### ISSUE THREAD

**Issue by copperwater** [NetHack/NetHack#378]
There are a couple of confusing rules with how vampshifter AI interacts with its hitpoints.
1. When killed, a vampshifter turns back into its normal form at full health.
2. A vampire will turn back into its normal form voluntarily at low health:
https://github.com/NetHack/NetHack/blob/cee29d33381c60c954bcafbd60ac9f4325354cee/src/mon.c#L3533-L3539
3. When a vampire voluntarily reverts to normal form, its proportion of HP is preserved:
https://github.com/NetHack/NetHack/blob/cee29d33381c60c954bcafbd60ac9f4325354cee/src/mon.c#L3969-L3980

The overall effect is that when fighting a vampire in a shifted form, it will usually go down for good in one hit if it decides to turn itself back voluntarily, but if it turns back involuntarily (through its shifted form being destroyed) it gets its full hitpoints back and will probably be harder to destroy. Because it takes its turn to change back, this often means that the player never actually has to contend with the vampire as a vampire - they just hit a vampire bat or wolf a couple times, it turns back into a vampire, and the player immediately destroys it. (I've even seen instances where players deliberately monitor the HP of a suspected vampshifter with a stethoscope, then hang back until it decides to revert and the player can easily kill it. Perhaps it indicates a different balance issue here that players have a lot less to fear from a shifted vampire in any form than an unshifted one...)

My proposal to fix this is that:
* Vampires killed in their shifted form return with a fixed fraction (say 50%) of their max HP.
* Vampires in their normal form and beneath 50% max HP are incapable of shifting forms.
* Having less than max HP in a shifted form translates to 50% of their max HP, with the other 50% proportional to the HP they had in their shifted form. E.g. if a vampire changes from a wolf at 8/20 HP back to its normal form at 60 max HP, it would get 30 (half of its max) plus 12 (8/20 scaled to the remaining 30). If they then changed back into a wolf while still at 42 HP, it would again have 8/20.

**Comment by jeremyn** [comment]
Is it on the table to just get rid of the vampire shapeshift ability entirely? For most of the game when I'm fighting vampires, their shapeshift form is entirely trivial, and having to punch through that first just tediously prolongs the fight. Maybe there's a brief window in the lower Dungeons of Doom where it adds a sliver of challenge, but for the big graveyards in the Valley and Moloch's Sanctum, not really, no. I guess it's intended as a sort of spooky "gotcha" but like, I'm not fooled when I see a fog cloud coming under a door at me in Vlad's, it just means I need to attack a few more times before moving on.

These hyper-optimized HP rules seem like overkill. I'm unsure who it is that's monitoring vampires with stethoscopes, maybe people running extremely underpowered conduct runs, or players on their first ascension clocking in at 100k moves?

**Comment by copperwater** [comment]
I agree that it is a problem that all the forms a vampire can take are less threatening than its default form, but removing that feature rather than fixing it is too extreme. I like vampire shifting because it's backed up by folklore/pop culture and helps differentiate them as a monster rather than just being another statblock in monst.c. Not to mention it was only introduced in 3.6; it's a baby feature in NetHack terms. Let's not jump the gun here.

Ideally I think I'd want stronger AI powering vampires' form choices: they would only turn into fog clouds when there's some passable obstacle (locked door, iron bars) between them and their destination, and then revert after that's no longer the case; they turn into wolves for speed in getting to their destination. I admit I can't think of any reason right now why they would turn into bats as long as they can fly in their default form, but that's the most classic of the three. Maybe it's just a form they can take when idling.

If they used their form changes situationally rather than as the default mode of combat, then it'd be more feasible to make them die for real when killed in a shifted form, which would make most of the HP calculations unnecessary. But overhauling AI to make those things a reality is out of scope for what I want to address in this issue, which is the problem of them retaining their low HP when voluntarily shifting back.

**Comment by jeremyn** [comment]
So I'm a solo player and I don't really know how vampire shapeshifting has been received by players generally. If you know, then I would unironically/genuinely appreciate you letting me know. In my own experience, by the time I'm fighting lots of vampires, I'm powerful enough to easily destroy their shapeshift form and the main tension is whether I can kill their vampire form before they level drain me. So the shapeshift form adds nothing to the fun for me, it's just some low level creature in the way.

But even though I can ascend if I concentrate, I'm a slow and mediocre player. Maybe these vampire micro-dynamics are really interesting to speedrunners or experts who enter Gehennom at level 14 with only half their ascension kit who really are using stethoscopes for a lot of their fights. If that's the case then I withdraw my complaints. However if most players are like me and shapeshifting just prolongs fights then it might be worth removing shapeshifting entirely which would make this issue irrelevant.

**Comment by copperwater** [comment]
I think that "vampire shapeshifting as it currently stands doesn't add much to the game" is a fairly common opinion which I share, but I am not aware of any calls to remove it other than this one. I'd prefer to make it add something interesting to the game instead.

Regardless, as I said, this issue is not aimed at fixing that problem, only at fixing a bug with the current AI logic.

**Comment by jeremyn** [comment]
When you're fixing bugs is a good time to decide if you can simplify the code overall by cutting stuff out that isn't needed. If vampshifting isn't wanted then this all becomes unnecessary.

Failing that, maybe we can simplify things without removing vampshifting entirely. How about: use the same logic as werecreatures except that they can never change out of vampire form, only into it. Done. Would that work? It's less "AI"-like but in my opinion NetHack's charm isn't in having "smart" creatures, it's allowing detailed interactions between a lot of simple things.

This could allow a quick player to kill an unfortunate shifted vampire before it can change into its more dangerous vampire form. There's a funny quote from the Discworld series (in *Witches Abroad*) where [Greebo the cat](https://discworld.fandom.com/wiki/Greebo) kills a vampire in bat form: 

> The bat squirmed under his claw. It seemed to Greebo's small cat brain that it was trying to change its shape, and he wasn't having any of that from a mouse with wings on.

If the player can get away with it, good for them.
